<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>

<h1 align="center"><font size="5">Supervised Machine Learning: Regression - Final Assignment</font></h1>


## Instructions:

In this Assignment, you will demonstrate the data regression skills you have learned by completing this course. You are expected to leverage a wide variety of tools, but also this report should focus on present findings, insights, and next steps. You may include some visuals from your code output, but this report is intended as a summary of your findings, not as a code review.

The grading will center around 5 main points:

1. Does the report include a section describing the data?
2. Does the report include a paragraph detailing the main objective(s) of this analysis?
3. Does the report include a section with variations of linear regression models and specifies which one is the model that best suits the main objective(s) of this analysis.
4. Does the report include a clear and well-presented section with key findings related to the main objective(s) of the analysis?
5. Does the report highlight possible flaws in the model and a plan of action to revisit this analysis with additional data or different predictive modeling techniques?




## Import the required libraries


The following required modules are pre-installed in the Skills Network Labs environment. However if you run this notebook commands in a different Jupyter environment (e.g. Watson Studio or Ananconda) you will need to install these libraries by removing the `#` sign before `!mamba` in the code cell below.


In [555]:
# All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented.
# !mamba install -qy pandas==1.3.4 numpy==1.21.4 seaborn==0.9.0 matplotlib==3.5.0 scikit-learn==0.20.1
# Note: If your environment doesn't support "!mamba install", use "!pip install"

In [556]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.decomposition import PCA
from sklearn.preprocessing import LabelEncoder

## Importing the Dataset


Before you begin, you will need to choose a data set that you feel passionate about. You can brainstorm with your peers about great public data sets using the discussion board in this module.


Read your chosen dataset into pandas dataframe:


In [557]:
data = pd.read_csv('https://raw.githubusercontent.com/Chibueze864/IBM_ML_CERTIFICATION_COURSE/main/Supervised%20Learning%20Projects/Final%20Project/all_billionaires_1997_2023.csv')
data.head(20)

year  month  rank net_worth     last_name            first_name  \
0   1997      7   NaN     2.0 B  Sophonpanich                Chatri   
1   1997      7   NaN     1.8 B     Adulyadej         King Bhumibol   
2   1998      7   NaN     3.3 B         Safra                Edmond   
3   1999      7   NaN     1.0 B       Hinduja  Srichand, Gopichand,   
4   1999      7   NaN     7.1 B       Schwarz                Dieter   
5   1999      7   NaN     1.3 B  Ardila Lulle                Carlos   
6   1999      7   NaN     2.9 B         Ebner                Martin   
7   1999      7   NaN     1.0 B       Cutrale                  Jose   
8   1999      7   NaN     1.1 B           Lee                 Peter   
9   1999      7   NaN     1.8 B         Kampf                 Serge   
10  1999      7   NaN     2.5 B         Safra                Edmond   
11  2000      7   NaN     1.0 B       Cutrale                  Jose   
12  2000      7   NaN     1.9 B         Ebner                Martin   
13  2000      7   NaN     4.5 B        Henkel                   NaN   
14  2000      7   NaN     2.9 B           Vik             Alexander   
15  2000      7   NaN     1.2 B          Noda              Masahiro   
16  2000      7   NaN     1.2 B        Kozuki              Kagemasa   
17  2000      7   NaN     2.1 B         Kampf                 Serge   
18  2000      7   0.0     1.4 B          Wada             Shigefumi   
19  2001      7   1.0    58.7 B         Gates               William   

                       full_name  birth_date   age gender  \
0   Chatri Sophonpanich & family  1934-02-28  73.0   Male   
1        King Bhumibol Adulyadej  1927-12-05  69.0   Male   
2                   Edmond Safra  1932-08-06  65.0   Male   
3   Srichand & Gopichand Hinduja  1935-11-28   NaN    NaN   
4                 Dieter Schwarz  1939-09-24  59.0   Male   
5            Carlos Ardila Lülle  1930-06-04  68.0   Male   
6                   Martin Ebner  1945-08-01  53.0   Male   
7              Jose Luis Cutrale  1946-09-17   NaN   Male   
8                     Lee family         NaN  71.0   Male   
9                    Serge Kampf  1934-01-01  65.0   Male   
10                  Edmond Safra  1932-08-06  66.0   Male   
11             Jose Luis Cutrale  1946-09-17   NaN   Male   
12                  Martin Ebner  1945-08-01  54.0   Male   
13              Christoph Henkel  1958-02-11   NaN    NaN   
14                 Alexander Vik  1955-02-19  45.0   Male   
15                 Masahiro Noda  1938-08-24  61.0   Male   
16               Kagemasa Kozuki  1940-11-12  59.0   Male   
17                   Serge Kampf  1934-01-01  66.0   Male   
18                Shigefumi Wada  1952-08-30  47.0   Male   
19                    Bill Gates  1955-10-28  45.0   Male   

   country_of_citizenship country_of_residence city_of_residence  \
0                Thailand             Thailand           Bangkok   
1                Thailand                  NaN               NaN   
2                 Lebanon              Lebanon               NaN   
3                   India       United Kingdom            London   
4                 Germany              Germany        Neckarsulm   
5                Colombia             Colombia            Bogota   
6             Switzerland          Switzerland             Wilen   
7                  Brazil               Brazil         Sao Paulo   
8               Hong Kong            Hong Kong         Hong Kong   
9                  France                  NaN               NaN   
10                Lebanon                  NaN               NaN   
11                 Brazil               Brazil         Sao Paulo   
12            Switzerland          Switzerland             Wilen   
13                Germany       United Kingdom            London   
14                 Norway               Monaco            Monaco   
15                  Japan                Japan             Tokyo   
16                  Japan                Japan             Tokyo   
17    

Once you have selected a data set, you will produce the deliverables listed below and submit them to one of your peers for review. Treat this exercise as an opportunity to produce analysis that are ready to highlight your analytical skills for a senior audience, for example, the Chief Data Officer, or the Head of Analytics at your company.
Sections required in your report:

*   Main objective of the analysis that specifies whether your model will be focused on prediction or interpretation.
*   Brief description of the data set you chose and a summary of its attributes.
*   Brief summary of data exploration and actions taken for data cleaning and feature engineering.
*   Summary of training at least three linear regression models which should be variations that cover using a simple  linear regression as a baseline, adding polynomial effects, and using a regularization regression. Preferably, all use the same training and test splits, or the same cross-validation method.
*  A paragraph explaining which of your regressions you recommend as a final model that best fits your needs in terms of accuracy and explainability.
*  Summary Key Findings and Insights, which walks your reader through the main drivers of your model and insights from your data derived from your linear regression model.
*  Suggestions for next steps in analyzing this data, which may include suggesting revisiting this model adding specific data features to achieve a better explanation or a better prediction.


# 1. About the Data


The "Forbes Billionaires Evolution" dataset provides a comprehensive examination of the financial growth and status of global billionaires from 1997 to 2023. It meticulously chronicles the transformations, both subtle and substantial, in the fortunes of these financial titans over nearly three decades.

The cover image used for this dataset is sourced from Forbes, with the illustration credited to Geoff J. Kim for Forbes.

Content:
This dataset offers a plethora of valuable information, including:

    Yearly rankings.
    Net worth details and evolution.
    Personal details like names, age, gender, and birth dates.
    Residence details spanning country, city, and citizenship.
    In-depth business category and industry classifications.
    Details about affiliations with various organizations, including positions held.
    Indicators like "self-made" and the current wealth status.


# 2. Objectives


My objective is to identify industries and regions with the most billionaire activity and growth. I would like to identify the sectors and geographical locations that produce the most wealth.
This task is focused on interpreting the data to find out why some industries produce more wealth than others

##3. Exploratory Data Analysis

Check the data types that make up this dataset along with its value distribution and null values


In [558]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31732 entries, 0 to 31731
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   year                      31732 non-null  int64  
 1   month                     31732 non-null  int64  
 2   rank                      31714 non-null  float64
 3   net_worth                 31732 non-null  object 
 4   last_name                 28043 non-null  object 
 5   first_name                27985 non-null  object 
 6   full_name                 31732 non-null  object 
 7   birth_date                27328 non-null  object 
 8   age                       31057 non-null  float64
 9   gender                    27903 non-null  object 
 10  country_of_citizenship    31723 non-null  object 
 11  country_of_residence      31030 non-null  object 
 12  city_of_residence         30797 non-null  object 
 13  business_category         25889 non-null  object 
 14  busine

So most of the data types in this data are categorical

In [559]:
data.describe()

year         month          rank           age
count  31732.000000  31732.000000  31714.000000  31057.000000
mean    2015.755956      3.364143    903.961374     62.835077
std        5.644194      0.604035    645.565108     14.175068
min     1997.000000      2.000000      0.000000      0.000000
25%     2012.000000      3.000000    366.000000     54.000000
50%     2017.000000      3.000000    782.000000     63.000000
75%     2021.000000      4.000000   1362.000000     73.000000
max     2023.000000      7.000000   2674.000000    101.000000

In [560]:
for col in data.columns:
    value_counts = data[col].value_counts().reset_index()
    value_counts.columns = [col, 'Counts']
    print(f"Value counts for {col}:\n{value_counts}\n")

Value counts for year:
    year  Counts
0   2021    2755
1   2022    2668
2   2023    2640
3   2018    2208
4   2019    2153
5   2020    2095
6   2017    2043
7   2015    1826
8   2016    1811
9   2014    1645
10  2013    1426
11  2012    1226
12  2011    1209
13  2010    1011
14  2008     908
15  2007     761
16  2009     738
17  2006     628
18  2005     530
19  2004     432
20  2001     335
21  2002     333
22  2003     332
23  2000       8
24  1999       8
25  1997       2
26  1998       1

Value counts for month:
   month  Counts
0      3   21208
1      4   10158
2      7     347
3      2      12
4      6       7

Value counts for rank:
       rank  Counts
0    1818.0     234
1    1415.0     183
2    1513.0     166
3    2524.0     150
4    1929.0     147
..      ...     ...
835   781.0       1
836   526.0       1
837  1152.0       1
838  1711.0       1
839     0.0       1

[840 rows x 2 columns]

Value counts for net_worth:
    net_worth  Counts
0       1.1 B    1738
1       1.2 B

Here I see that ordinal encoding will not work for most of the categorical columns because the sheer number of categories.

Check for null values

In [561]:
data.isna().sum()

year                            0
month                           0
rank                           18
net_worth                       0
last_name                    3689
first_name                   3747
full_name                       0
birth_date                   4404
age                           675
gender                       3829
country_of_citizenship          9
country_of_residence          702
city_of_residence             935
business_category            5843
business_industries           990
organization_name           27344
position_in_organization    27040
self_made                    3689
wealth_status                7146
dtype: int64

Select the apprioprate features for our task. As I stated above my goal is to identify industries and regions with the most billionaire activity and growth.

We are only concerned about the factors that account for wealth generation. We want to identify the most important features that push the model's insights in the data

In [563]:
data_interest = data[["net_worth","business_category","business_industries","country_of_residence","rank"]]

Get numeric part of *net_worth* feature. For every number extracted, put it in a new column called *net_worth_billions* that will replace the *net_worth* column

Fill Null Values

target encode all columns using the year column as the category

Fill the null values in the dataframe

In [564]:
def prepare_data(data_interest):

  data_interest['net_worth_billions'] = data_interest['net_worth'].apply(lambda x: ''.join(char for char in x if char.isdigit()))
  data_interest = data_interest.drop("net_worth", axis="columns")
  for col in data_interest.columns:
    # Calculate the most common value in the column
    most_common_value = data_interest[col].mode()[0]

    # Replace missing values with the most common value
    data_interest[col] = data_interest[col].fillna(most_common_value)
  for col in data_interest.columns:
      if col != "net_worth_billions":
            # Calculate mean target value for each category
            target_means = data_interest.groupby(col)["net_worth_billions"].median()

            # Map target means to the categorical column
            data_interest[col] = data_interest[col].map(target_means)

  return data_interest

data_interest.nunique()


net_worth               541
business_category        41
business_industries      67
country_of_residence     94
rank                    840
dtype: int64

In [565]:
data_to_be_used = prepare_data(data_interest)
data_to_be_used.head()

<ipython-input-564-cf0aab864ac7>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_interest['net_worth_billions'] = data_interest['net_worth'].apply(lambda x: ''.join(char for char in x if char.isdigit()))


business_category  business_industries  country_of_residence  rank  \
0               22.0                 21.0                  21.0  12.0   
1               22.0                 20.0                  24.0  12.0   
2               22.0                 20.0                  25.0  12.0   
3               22.0                 20.0                  25.0  12.0   
4               25.0                 25.0                  27.0  12.0   

  net_worth_billions  
0                 20  
1                 18  
2                 33  
3                 10  
4                 71

In [566]:
# Calculate percentage of null values for each feature
null_percentages = (data_to_be_used.isna().sum() / len(data)) * 100

# Print the results
print("Percentage of null values for each feature: \n {}".format(null_percentages) )



Percentage of null values for each feature: 
 business_category       0.0
business_industries     0.0
country_of_residence    0.0
rank                    0.0
net_worth_billions      0.0
dtype: float64


In [567]:
data_to_be_used.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31732 entries, 0 to 31731
Data columns (total 5 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   business_category     31732 non-null  float64
 1   business_industries   31732 non-null  float64
 2   country_of_residence  31732 non-null  float64
 3   rank                  31732 non-null  float64
 4   net_worth_billions    31732 non-null  object 
dtypes: float64(4), object(1)
memory usage: 1.2+ MB


# 3. Linear Regression Models


Pipeline for each model

In [568]:
X = data_to_be_used.drop("net_worth_billions", axis="columns")
y = data_to_be_used["net_worth_billions"]


In [569]:
# Create a pipeline for Linear Regression
linear_pipeline = Pipeline([
    ('polynomial_features', PolynomialFeatures(degree=2)),

    ('scaler', StandardScaler()),
    ('linear_regression', LinearRegression())
])


param_grid = {
    'linear_regression__fit_intercept': [True, False],  # Include or exclude intercept
    'linear_regression__n_jobs': [1,2,3,4,5]       # Normalize features
}

# Perform grid search and cross-validation
grid_search = GridSearchCV(linear_pipeline, param_grid, cv=5)
grid_search.fit(X, y)


# Print best parameters and score
print("Best Parameters:", grid_search.best_params_)
# print("Best Cross-Validation Score:", grid_search.best_score_)
print("Feature Model:", grid_search.best_estimator_)

best_linear = grid_search.best_estimator_["linear_regression"]
best_linear.fit(X,y)
best_linear.coef_
best_linear.score(X,y)


Best Parameters: {'linear_regression__fit_intercept': True, 'linear_regression__n_jobs': 1}
Feature Model: Pipeline(steps=[('polynomial_features', PolynomialFeatures()),
                ('scaler', StandardScaler()),
                ('linear_regression', LinearRegression(n_jobs=1))])


0.7085263255842932

In [570]:
# Create a pipeline for Ridge Regression
ridge_pipeline = Pipeline([
    ('polynomial_features', PolynomialFeatures(degree=2)),
    ('scaler', StandardScaler()),
    ('ridge_regression', Ridge())
])


param_grid = {
    'ridge_regression__alpha': [0.1, 1.0],

}

# Perform grid search and cross-validation
grid_search = GridSearchCV(ridge_pipeline, param_grid, cv=5)
grid_search.fit(X, y)

# Print best parameters and score
print("Best Parameters:", grid_search.best_params_)
# print("Best Cross-Validation Score:", grid_search.best_score_)
print("Feature Model:", grid_search.best_estimator_)

best_ridge = grid_search.best_estimator_["ridge_regression"]
best_ridge.fit(X,y)
best_ridge.coef_
best_ridge.score(X,y)

Best Parameters: {'ridge_regression__alpha': 1.0}
Feature Model: Pipeline(steps=[('polynomial_features', PolynomialFeatures()),
                ('scaler', StandardScaler()), ('ridge_regression', Ridge())])


0.7085263255842679

In [571]:
# Create a pipeline for Lasso Regression
lasso_pipeline = Pipeline([
    ('polynomial_features', PolynomialFeatures(degree=2)),
    ('scaler', StandardScaler()),
    ('lasso_regression', Lasso(alpha=1.0))
])

param_grid = {
    'lasso_regression__alpha': [0.1, 1.0],

}

# Perform grid search and cross-validation
grid_search = GridSearchCV(lasso_pipeline, param_grid, cv=5)
grid_search.fit(X, y)

# Print best parameters and score
print("Best Parameters:", grid_search.best_params_)
# print("Best Cross-Validation Score:", grid_search.best_score_)
print("Feature Model:", grid_search.best_estimator_)

best_lasso = grid_search.best_estimator_["lasso_regression"]
best_lasso.fit(X,y)
best_lasso.coef_
best_lasso.score(X,y)

Best Parameters: {'lasso_regression__alpha': 1.0}
Feature Model: Pipeline(steps=[('polynomial_features', PolynomialFeatures()),
                ('scaler', StandardScaler()), ('lasso_regression', Lasso())])


0.708391066590426

In [572]:

# Create a pipeline for Elastic Net with Polynomial Features and Standard Scaler
elastic_net_pipeline = Pipeline([

    ('polynomial_features', PolynomialFeatures(degree=2)),

    ('scaler', StandardScaler() ),
    ('elastic_net', ElasticNet()), # No hyperparameters set here

])

# Parameters for grid search
param_grid = {
    'elastic_net__alpha': [ 0.01, 1.0],
    'elastic_net__l1_ratio':  [ 0.01, 1.0],
}

# Perform grid search and cross-validation
grid_search = GridSearchCV(elastic_net_pipeline, param_grid, cv=5)
grid_search.fit(X, y)

# Print best parameters and score
print("Best Parameters:", grid_search.best_params_)
# print("Best Cross-Validation Score:", grid_search.best_score_)
print("Feature Model:", grid_search.best_estimator_)

best_enet = grid_search.best_estimator_["elastic_net"]
best_enet.fit(X,y)
best_enet.coef_
best_enet.score(X,y)




/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.681e+06, tolerance: 1.441e+04
  model = cd_fast.enet_coordinate_descent(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.045e+05, tolerance: 1.427e+04
  model = cd_fast.enet_coordinate_descent(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.408e+05, tolerance: 1.222e

Best Parameters: {'elastic_net__alpha': 1.0, 'elastic_net__l1_ratio': 0.01}
Feature Model: Pipeline(steps=[('polynomial_features', PolynomialFeatures()),
                ('scaler', StandardScaler()),
                ('elastic_net', ElasticNet(l1_ratio=0.01))])


0.7085110444820251

In [576]:
print("ElasticNet Most important factors are: ", list(zip(X.columns, abs(best_enet.coef_))))
print("Lasso Most important factors are: ", list(zip(X.columns, abs(best_lasso.coef_))))
print("Ridge Most important factors are: ", list(zip(X.columns, abs(best_ridge.coef_))))
print("Vanilla regression Most important factors are: ", list(zip(X.columns, abs(best_linear.coef_))))



ElasticNet Most important factors are:  [('business_category', 0.3669878045019928), ('business_industries', 0.29171866624044457), ('country_of_residence', 0.2652041626358142), ('rank', 1.1336814326480917)]
Lasso Most important factors are:  [('business_category', 0.13717922426501783), ('business_industries', 0.09018367785579526), ('country_of_residence', 0.22316427151835672), ('rank', 1.1329581079214066)]
Ridge Most important factors are:  [('business_category', 0.4882399541114604), ('business_industries', 0.39198630648939237), ('country_of_residence', 0.28180673321442645), ('rank', 1.134380197496274)]
Vanilla regression Most important factors are:  [('business_category', 0.4882449520217471), ('business_industries', 0.39199044396446253), ('country_of_residence', 0.28180727405168626), ('rank', 1.1343802211336305)]


# 4. Insights and key findings


From my findings, I saw that the factors such as
1. Business Category
2. Business Industries
3. Country of residence

Are correlated to the Net Worth about 71% of the time. The 29% unexplained variance can largely be accounted to the missing data in the data set.

This makes sense because the forces in the industry play a huge role on how much wealth the people can acquire. Also the business is at the mercy of the policies and resources available in that country


# 5. Next Steps


Yes there is a flaw in the dataset that limits the amount of insights. A richer dataset would improve the quality of predicitions

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
